File: rust/src/logger/processors.rs
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.
//
//! Logger processor chain — runs on every LogEvent before emission.
//!
//! Mirrors the Python `processors.py` pipeline:
//! 1. DARS extraction — extract domain/action/resource/status from Event
//! 2. Logger name injection — insert target as logger_name field if absent
//! 3. Harden input — truncate values, strip control chars, limit attr count
//! 4. Error fingerprint — compute stable hash for error-level events
//! 5. PII sanitization — redact/hash sensitive fields via `pii.rs`
//! 6. Schema enforcement — validate event name format when strict mode is on
//!
//! Called from `log_event()` after the event is built and before emission.

use std::collections::BTreeMap;

use serde_json::Value;

use crate::config::TelemetryConfig;
use crate::fingerprint::compute_error_fingerprint;
use crate::harden::{clean_key, cleaned_key_claims_slot, harden_value, HardenLimits};
use crate::pii::{detect_secret_in_string, redact_secret_spans, sanitize_payload, REDACTED_SENTINEL};
use crate::runtime::get_runtime_config;
use crate::schema::{event_name, get_strict_schema, validate_required_keys};

use super::LogEvent;

fn runtime_config_or_env() -> Option<TelemetryConfig> {
    match get_runtime_config() {
        Some(cfg) => Some(cfg),
        None => TelemetryConfig::from_env().ok(),
    }
}

fn first_context_string<'a>(event: &'a LogEvent, keys: &[&str]) -> Option<&'a str> {
    for key in keys {
        if let Some(Value::String(value)) = event.context.get(*key) {
            return Some(value.as_str());
        }
    }
    None
}

fn runtime_schema_error(event: &LogEvent) -> Option<String> {
    let cfg = get_runtime_config()?;
    match validate_required_keys(&event.context, &cfg.event_schema.required_keys) {
        Ok(()) => None,
        Err(err) => Some(err.message),
    }
}

fn is_priority_key(key: &str) -> bool {
    matches!(
        key,
        "service"
            | "env"
            | "version"
            | "trace_id"
            | "span_id"
            | "session_id"
            | "logger_name"
            | "domain"
            | "action"
            | "resource"
            | "status"
            | "error_fingerprint"
    )
}

/// Run the full processor chain on a LogEvent in place.
///
/// The order matches the Python processor chain (except sampling/consent
/// which run before this point in `log_event()`).
pub(super) fn process_event(event: &mut LogEvent) {
    let cfg = runtime_config_or_env();
    let pii_max_depth = cfg.as_ref().map_or(8, |c| c.pii_max_depth);
    let limits = HardenLimits {
        max_value_length: cfg
            .as_ref()
            .map_or(1024, |c| c.security.max_attr_value_length),
        max_attr_count: cfg.as_ref().map_or(64, |c| c.security.max_attr_count),
        max_depth: cfg.as_ref().map_or(8, |c| c.security.max_nesting_depth),
    };

    // 1. DARS extraction (only when Event metadata is attached)
    extract_dars_fields(event);

    // 2. Logger name injection — inserts target as logger_name if not already set
    inject_logger_name(event);

    // 3. Harden input
    harden_input(event, limits);

    // 4. Error fingerprint
    add_error_fingerprint(event);

    // 5. PII sanitization
    sanitize_context(event, pii_max_depth);

    // 6. Schema enforcement (validate event name when strict mode is on)
    // Annotates with _schema_error instead of dropping — cross-language
    // standard: all four languages (Python/TypeScript/Go/Rust) annotate and emit.
    enforce_schema(event);
}

/// Extract DARS (Domain.Action.Resource.Status) fields from an attached
/// Event into the context map. Only fires when `event.event_metadata` is
/// populated (i.e., the caller used `Logger::info_event(&event)` rather
/// than `Logger::info("message")`).
fn extract_dars_fields(event: &mut LogEvent) {
    let meta = match event.event_metadata.as_ref() {
        Some(meta) => meta,
        None => return,
    };
    event
        .context
        .insert("domain".to_string(), Value::String(meta.domain.clone()));
    event
        .context
        .insert("action".to_string(), Value::String(meta.action.clone()));
    if let Some(resource) = meta.resource.as_ref() {
        event
            .context
            .insert("resource".to_string(), Value::String(resource.clone()));
    }
    event
        .context
        .insert("status".to_string(), Value::String(meta.status.clone()));
}

/// Inject the logger's target name as `logger_name` into the context.
/// Only sets the field when it is absent — caller-provided values are
/// preserved. Matches Python's `inject_logger_name` processor.
fn inject_logger_name(event: &mut LogEvent) {
    if event.target.is_empty() {
        return;
    }
    if event.context.contains_key("logger_name") {
        return;
    }
    event.context.insert(
        "logger_name".to_string(),
        Value::String(event.target.clone()),
    );
}

/// Bound every context value, then cap how many of them survive.
///
/// The values are hardened recursively — depth, width and length at every
/// level — so nothing downstream (PII traversal, receipt canonicalization,
/// renderers, the OTel bridge) has to defend itself against an unbounded
/// structure. The top-level key cap is separate from the one inside
/// [`harden_value`] because it is not a plain first-N: dropping `trace_id` or
/// `service` to make room for a caller's forty-first keyword argument would
/// cost more than it saves.
fn harden_input(event: &mut LogEvent, limits: HardenLimits) {
    for value in event.context.values_mut() {
        // Depth 1: the context map itself is the depth-0 composite.
        *value = harden_value(value, limits, 1);
    }
    cap_attr_count(event, limits.max_attr_count);
    // Top-level keys are hardened as well as values, after the cap — the same
    // order as Python's `harden_input`. [`harden_value`] cleans the keys of
    // *nested* maps, but the context map's own keys — the ones the pretty
    // renderer emits bare as `key=value` — never pass through it, so a key
    // containing "\n2026-08-10 [error] ..." would forge a second,
    // attacker-controlled log line. Mirrors Python's `_harden_keys`.
    harden_keys(&mut event.context);
}

/// Rebuild the context under cleaned keys, resolving collisions safely.
///
/// Cleaning is many-to-one, and the collision policy
/// ([`cleaned_key_claims_slot`]) is what keeps a forged `"trace_i\x00d"`
/// payload key from replacing the genuine `trace_id` and correlating the
/// record to an attacker-chosen trace.
fn harden_keys(context: &mut BTreeMap<String, Value>) {
    use std::collections::BTreeSet;
    let original = std::mem::take(context);
    let mut verbatim: BTreeSet<String> = BTreeSet::new();
    for (key, value) in original {
        let name = clean_key(&key);
        let untouched = name == key;
        if !cleaned_key_claims_slot(
            context.contains_key(&name),
            untouched,
            verbatim.contains(&name),
        ) {
            continue;
        }
        context.insert(name.clone(), value);
        if untouched {
            verbatim.insert(name);
        }
    }
}

/// Cap how many top-level keys survive, keeping priority fields first.
fn cap_attr_count(event: &mut LogEvent, max_attr_count: usize) {
    if max_attr_count == 0 {
        return;
    }
    if event.context.len() <= max_attr_count {
        return;
    }

    use std::collections::BTreeSet;
    let mut keep = BTreeSet::new();
    let priority_keys: Vec<String> = event.context.keys().cloned().collect();
    for key in &priority_keys {
        if is_priority_key(key) {
            keep.insert(key.clone());
        }
    }

    let candidate_keys: Vec<String> = event.context.keys().cloned().collect();
    for key in &candidate_keys {
        if keep.len() >= max_attr_count {
            break;
        }
        if is_priority_key(key) {
            continue;
        }
        keep.insert(key.clone());
    }
    let original_context = std::mem::take(&mut event.context);
    let mut retained = BTreeMap::new();
    for (key, value) in original_context {
        if keep.contains(&key) {
            retained.insert(key, value);
        }
    }
    event.context = retained;
}

/// Compute a stable error fingerprint for ERROR/CRITICAL-level events
/// that carry an `error` or `error_type` key in context. The fingerprint
/// is a 12-char hex SHA256 prefix, stable across minor stack variations.
fn add_error_fingerprint(event: &mut LogEvent) {
    const ERROR_LEVELS: &[&str] = &["ERROR", "CRITICAL", "FATAL"];
    if !ERROR_LEVELS.contains(&event.level.as_str()) {
        return;
    }
    let error_name = first_context_string(event, &["error", "error_type", "exception"]);
    // Only add fingerprint when error/exception attributes exist in the
    // context — matching Python/TypeScript/Go semantics. Plain error-level
    // messages without error metadata do not get fingerprinted.
    let error_name = match error_name {
        Some(error_name) => error_name,
        None => return,
    };
    let stack = first_context_string(event, &["stack", "stacktrace"]);
    let fingerprint = compute_error_fingerprint(error_name, stack);
    event
        .context
        .insert("error_fingerprint".to_string(), Value::String(fingerprint));
}

/// Sanitize PII/secrets in the context map and scrub the free-form message
/// string. Message is checked directly (not via the map-based engine) so
/// `Path: ["*"]` rules can't match a sentinel key. Mirrors Python/Go.
fn sanitize_context(event: &mut LogEvent, max_depth: usize) {
    let message_has_secret = detect_secret_in_string(&event.message);
    if message_has_secret {
        event.message = redact_secret_spans(&event.message);
    }
    if event.context.is_empty() {
        return;
    }
    let payload = Value::Object(event.context.clone().into_iter().collect());
    let cleaned = sanitize_payload(&payload, true, max_depth);
    let object = cleaned
        .as_object()
        .expect("sanitize_payload preserves object shape")
        .clone();
    event.context = object.into_iter().collect();
}

/// When strict schema mode is on, validate the event message as a
/// dot-joined event name. Invalid names get a `_schema_error` context
/// field — the event is always emitted (never dropped), so telemetry
/// is never lost. Cross-language standard: all four languages annotate
/// and emit rather than drop.
fn enforce_schema(event: &mut LogEvent) {
    if let Some(message) = runtime_schema_error(event) {
        event
            .context
            .insert("_schema_error".to_string(), Value::String(message));
        return;
    }
    if !get_strict_schema() {
        return;
    }
    let segments: Vec<&str> = event.message.split('.').collect();
    match event_name(&segments) {
        Ok(_) => {}
        Err(_) => {
            event.context.insert(
                "_schema_error".to_string(),
                Value::String(format!(
                    "event name {:?} does not match strict schema",
                    event.message
                )),
            );
        }
    }
}

#[cfg(test)]
#[path = "processors_tests.rs"]
mod tests;

#[cfg(test)]
#[path = "processors_message_pii_tests.rs"]
mod message_pii_tests;

#[cfg(test)]
#[path = "processors_edge_tests.rs"]
mod edge_tests;

#[cfg(test)]
#[path = "processors_key_hardening_tests.rs"]
mod key_hardening_tests;

File: src/provide/telemetry/logger/processors.py
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

"""Structlog processors."""

from __future__ import annotations

import hashlib
import logging
import re
import sys
import traceback
import types
from typing import Any

import structlog

from provide.telemetry.config import TelemetryConfig
from provide.telemetry.logger.context import get_context
from provide.telemetry.schema.events import EventSchemaError, validate_event_name, validate_required_keys
from provide.telemetry.tracing.context import get_span_id, get_trace_id


def _get_active_config() -> Any | None:
    """Return the active runtime config without eagerly loading the runtime module."""
    runtime = sys.modules.get("provide.telemetry.runtime")
    if runtime is None:
        return None
    return getattr(
        runtime, "_active_config", None
    )  # pragma: no mutate — sentinel getattr default; "" would be truthy-compatible but semantically identical here


_CONTROL_CHAR_RE = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")
# Keys are rendered bare by the console renderer, so unlike values they cannot
# keep TAB/LF/CR: any of the three splits or misaligns the rendered line.
_CONTROL_CHAR_KEY_RE = re.compile(r"[\x00-\x1f\x7f]")


def _clean_key(key: object) -> str:
    """Strip every control character from an attribute key."""
    return _CONTROL_CHAR_KEY_RE.sub("", str(key))


def _harden_keys(event_dict: dict[str, Any]) -> dict[str, Any]:
    """Rebuild *event_dict* under cleaned keys, resolving collisions safely.

    Cleaning is many-to-one: ``"trace_i\\x00d"`` and ``"trace_id"`` both come out
    as ``"trace_id"``. A plain dict comprehension lets the later one win, and
    structlog's ``merge_contextvars`` always inserts contextvar-bound fields
    (``trace_id``, ``span_id``, ``request_id``, ``session_id``) *before* the
    caller's keyword arguments — so a request payload forwarded as
    ``logger.info(event, **payload)`` with a control-character key could replace
    the real bound value and correlate the record to an attacker-chosen trace.

    So a key that needed cleaning never displaces one already present, and a key
    that needed no cleaning always wins over a sanitized one that got there
    first. Two sanitized keys that collide keep the first, which is arbitrary but
    at least does not lose a genuine field.
    """
    resolved: dict[str, Any] = {}
    verbatim: set[str] = set()
    for key, value in event_dict.items():
        text = str(key)
        name = _clean_key(text)
        untouched = name == text
        if name in resolved and not (untouched and name not in verbatim):
            continue
        # Re-assigning an existing name keeps its original insertion position,
        # so a verbatim key reclaiming its slot does not reorder the record.
        resolved[name] = value
        if untouched:
            verbatim.add(name)
    return resolved


# Keys that must survive harden_input truncation regardless of insertion order.
# These are structlog/telemetry control fields; losing them silently corrupts
# routing, filtering, and trace correlation downstream.
_HARDEN_PRIORITY_KEYS: frozenset[str] = frozenset(
    {"event", "level", "timestamp", "trace_id", "span_id", "logger", "logger_name"}
)

TRACE_LEVEL = 5

# Fast lowercase level → numeric lookup (avoids normalize + getLevelName per message)
_FAST_LEVEL_LOOKUP: dict[str, int] = {
    "critical": logging.CRITICAL,
    "error": logging.ERROR,
    "warning": logging.WARNING,
    "info": logging.INFO,
    "debug": logging.DEBUG,
    "trace": TRACE_LEVEL,
}


def inject_das_fields(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    """Extract DA(R)S fields from an :class:`Event` instance into the log record."""
    from provide.telemetry.schema.events import Event

    ev = event_dict.get("event")
    if isinstance(ev, Event):
        event_dict["domain"] = ev.domain
        event_dict["action"] = ev.action
        if ev.resource is not None:
            event_dict["resource"] = ev.resource
        event_dict["status"] = ev.status
        event_dict["event"] = str(ev)
    return event_dict


def merge_runtime_context(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    event_dict.update(get_context())
    trace_id = get_trace_id()
    span_id = get_span_id()
    if trace_id is not None:
        event_dict["trace_id"] = trace_id
    if span_id is not None:
        event_dict["span_id"] = span_id
    return event_dict


def inject_logger_name(logger: Any, _: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    """Ensure structured JSON output exposes the logger name canonically."""
    name = event_dict.get("logger_name") or event_dict.get("logger")
    if name is None:
        name = getattr(logger, "name", None)
    if name:
        event_dict["logger_name"] = str(name)
    return event_dict


def _compute_error_fingerprint(exc_type: str, tb: types.TracebackType | None) -> str:
    """Generate a stable 12-char hex fingerprint from exception type + top 3 frames."""
    parts = [exc_type.lower()]
    if tb is not None:
        for frame in traceback.extract_tb(tb)[-3:]:
            # No maxsplit, and no `# pragma: no mutate`. Taking [-1] made the
            # maxsplit unobservable, so it only ever produced equivalent mutants
            # — but the bare pragma needed to silence them also hid the
            # separator normalisation and the index, which are not equivalent.
            # Without the argument, split/rsplit are indistinguishable here and
            # every mutation mutmut still generates changes the fingerprint.
            segments = frame.filename.replace("\\", "/").rsplit("/")
            leaf = segments[-1]
            basename = leaf.rsplit(".", 1)[0].lower()
            func = (frame.name or "").lower()
            parts.append(f"{basename}:{func}")
    fingerprint_bytes = ":".join(parts).encode("utf-8")  # pragma: no mutate — codec alias
    return hashlib.sha256(
        fingerprint_bytes
    ).hexdigest()[
        :12
    ]  # pragma: no mutate — 12-char truncation is a deliberate fingerprint-size choice; exact value asserted by fingerprint tests


def add_error_fingerprint(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    """Structlog processor: add stable error_fingerprint on error events."""
    exc_info = event_dict.get("exc_info")
    if exc_info is True:
        exc_info = sys.exc_info()
    if isinstance(exc_info, tuple) and len(exc_info) == 3 and exc_info[1] is not None:
        exc_type_name = type(exc_info[1]).__name__
        event_dict["error_fingerprint"] = _compute_error_fingerprint(exc_type_name, exc_info[2])
        return event_dict
    if isinstance(exc_info, BaseException):
        event_dict["error_fingerprint"] = _compute_error_fingerprint(type(exc_info).__name__, exc_info.__traceback__)
        return event_dict
    exc_name = event_dict.get("exc_name") or event_dict.get("exception")
    if exc_name:
        event_dict["error_fingerprint"] = _compute_error_fingerprint(str(exc_name), None)
    return event_dict


def harden_input(max_value_length: int, max_attr_count: int, max_depth: int) -> Any:
    """Structlog processor: truncate values, strip control chars, limit attributes."""

    def _processor(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
        live = _get_active_config()
        _max_value_length = live.security.max_attr_value_length if live is not None else max_value_length
        _max_attr_count = live.security.max_attr_count if live is not None else max_attr_count
        _max_depth = live.security.max_nesting_depth if live is not None else max_depth

        def _clean_value(value: object, depth: int) -> object:
            if isinstance(value, str):
                cleaned = _CONTROL_CHAR_RE.sub("", value)
                # `>=` is equivalent: slicing a value already at the limit to
                # [:limit] returns the identical string.
                too_long = len(cleaned) > _max_value_length  # pragma: no mutate
                if too_long:
                    return cleaned[:_max_value_length]
                return cleaned
            if isinstance(value, dict) and depth < _max_depth:
                return {k: _clean_value(v, depth + 1) for k, v in value.items()}
            if isinstance(value, list) and depth < _max_depth:
                return [
                    _clean_value(item, depth + 1) for item in value
                ]  # pragma: no mutate — list-comp traversal; element ordering asserted by nested-list tests
            return value

        # `>=` is equivalent: at exactly the limit the rebuild keeps every key,
        # and dict equality ignores the reordering it introduces.
        over_budget = _max_attr_count > 0 and len(event_dict) > _max_attr_count  # pragma: no mutate
        if over_budget:
            # Preserve control/telemetry fields first, then fill with user payload.
            # Simple first-N truncation would silently drop level, trace_id, etc.
            # when callers pass many keyword arguments.
            priority = {k: event_dict[k] for k in _HARDEN_PRIORITY_KEYS if k in event_dict}
            remaining = max(0, _max_attr_count - len(priority))
            user_keys = [k for k in event_dict if k not in _HARDEN_PRIORITY_KEYS]
            event_dict = {**priority, **{k: event_dict[k] for k in user_keys[:remaining]}}
        # Keys are hardened as well as values. The console renderer quotes values
        # but emits keys bare, so a control character in a key — which can reach
        # here from an untrusted W3C baggage header — would forge a log record.
        # parse_baggage rejects such keys at the boundary; this is the second line
        # of defence, and also covers keys a caller passes directly. _harden_keys
        # rather than a plain comprehension because cleaning is many-to-one: see
        # its docstring for the trace_id-shadowing vector that opens up otherwise.
        return {k: _clean_value(v, 0) for k, v in _harden_keys(event_dict).items()}

    return _processor


def add_standard_fields(config: TelemetryConfig) -> Any:
    def _processor(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
        event_dict.setdefault("service", config.service_name)
        event_dict.setdefault("env", config.environment)
        event_dict.setdefault("version", config.version)
        live = _get_active_config()
        include_error_taxonomy = (
            live.slo.include_error_taxonomy if live is not None else config.slo.include_error_taxonomy
        )
        if include_error_taxonomy and "error_type" not in event_dict and "exc_name" in event_dict:
            from provide.telemetry.slo import classify_error  # lazy: avoid loading metrics at logging config time

            status_code = event_dict.get("status_code")
            typed_status = status_code if isinstance(status_code, int) else None
            event_dict.update(classify_error(str(event_dict["exc_name"]), typed_status))
        return event_dict

    return _processor


_BACKPRESSURE_TICKET_KEY = "__provide_telemetry_backpressure_ticket__"


def apply_sampling(_: Any, method_name: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    from provide.telemetry.backpressure import try_acquire
    from provide.telemetry.consent import should_allow
    from provide.telemetry.health import increment_emitted
    from provide.telemetry.sampling import should_sample

    if not should_allow("logs", method_name):
        raise structlog.DropEvent()
    event_name = str(
        event_dict.get("event", "")
    )  # pragma: no mutate — empty-string default is fed into sampler which treats "" as unnamed; cosmetic vs any empty sentinel
    if not should_sample("logs", event_name):
        raise structlog.DropEvent()
    ticket = try_acquire("logs")
    if ticket is None:
        raise structlog.DropEvent()  # backpressure full; dropped counter already incremented
    increment_emitted("logs")
    # Stash the ticket. The final renderer processor moves it onto the
    # LogRecord via logging `extra`, and the stdlib handler boundary releases
    # it once all configured child handlers return from emit().
    event_dict[_BACKPRESSURE_TICKET_KEY] = ticket
    return event_dict


def render_with_backpressure_extra(renderer: Any) -> Any:
    """Render the event and attach any stashed ticket to stdlib logging `extra`."""

    def _processor(logger: Any, method_name: str, event_dict: dict[str, Any]) -> tuple[tuple[Any, ...], dict[str, Any]]:
        ticket = event_dict.pop(_BACKPRESSURE_TICKET_KEY, None)
        rendered = renderer(logger, method_name, event_dict)
        if ticket is None:
            return (rendered,), {}
        return (rendered,), {"extra": {_BACKPRESSURE_TICKET_KEY: ticket}}

    return _processor


def enforce_event_schema(config: TelemetryConfig) -> Any:
    def _processor(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
        live = _get_active_config()
        live_strict = live.strict_schema if live is not None else config.strict_schema
        live_event_schema = live.event_schema if live is not None else config.event_schema
        strict_event_name = True if live_strict else live_event_schema.strict_event_name
        required_keys = live_event_schema.required_keys
        event = str(event_dict.get("event", ""))
        try:
            validate_event_name(event, strict_event_name=strict_event_name)
            validate_required_keys(event_dict, required_keys)
        except EventSchemaError as exc:
            # Annotate instead of dropping — preserves telemetry while flagging
            # the schema violation.  Consumers can filter on _schema_error.
            # This is the cross-language standard (Rust/TypeScript/Go match).
            event_dict["_schema_error"] = str(exc)
        return event_dict

    return _processor


def sanitize_sensitive_fields(
    enabled: bool, max_depth: int = 8
) -> Any:  # pragma: no mutate — default max_depth=8 is overridden by live runtime config at every call; default value is only cosmetic
    from provide.telemetry.pii import sanitize_payload

    def _processor(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
        live = _get_active_config()
        _max_depth = live.pii_max_depth if live is not None else max_depth
        return sanitize_payload(event_dict, enabled, max_depth=_max_depth)

    return _processor


class _LevelFilter:
    """Per-module log level filter.

    FilteringBoundLogger handles the default level at zero cost.  This
    processor handles **module-level overrides** — e.g. ``asyncio=WARNING``
    while the default is ``INFO``.  It drops events whose level is below
    the threshold for their module (matched by longest-prefix).

    Placed late in the processor chain so enrichment processors run first.
    """

    __slots__ = ("_default_numeric", "_module_numerics", "_sorted_prefixes")

    def __init__(self, default_level: str, module_levels: dict[str, str]) -> None:
        self._default_numeric = _FAST_LEVEL_LOOKUP.get(default_level.lower(), logging.INFO)
        self._module_numerics: dict[str, int] = {
            module: _FAST_LEVEL_LOOKUP.get(lvl.lower(), logging.INFO) for module, lvl in module_levels.items()
        }
        # Longest prefix first for correct matching
        self._sorted_prefixes = sorted(self._module_numerics.keys(), key=lambda k: len(k), reverse=True)

    def __call__(self, _: Any, method_name: str, event_dict: dict[str, Any]) -> dict[str, Any]:
        logger_name: str = event_dict.get("logger_name", event_dict.get("logger", ""))
        event_level = _FAST_LEVEL_LOOKUP.get(event_dict.get("level", method_name).lower(), logging.INFO)

        threshold = self._default_numeric
        for prefix in self._sorted_prefixes:
            if prefix == "" or logger_name == prefix or logger_name.startswith(prefix + "."):
                threshold = self._module_numerics[prefix]
                break

        if event_level < threshold:
            raise structlog.DropEvent()
        return event_dict


def make_level_filter(default_level: str, module_levels: dict[str, str]) -> _LevelFilter:
    """Create a _LevelFilter for per-module log level overrides."""
    return _LevelFilter(default_level, module_levels)


def rename_event_to_message(_: Any, __: str, event_dict: dict[str, Any]) -> dict[str, Any]:
    """Rename structlog's 'event' key to canonical 'message' before JSON rendering.

    All four language loggers must emit 'message' as the message field.  structlog
    uses 'event' internally; this processor is inserted immediately before the
    JSONRenderer so the rename only affects the serialised output — all upstream
    processors (schema enforcement, PII sanitization, harden_input, etc.) still
    operate on 'event' as normal.
    """
    if "event" in event_dict:
        event_dict["message"] = event_dict.pop("event")
    return event_dict
